### PR TITLE
Fixed issues related to Gradle 6 and Android Studio 4.0 gradle plugin

### PR DESCRIPTION
**Pull request by fmoda3**
Android Studio 4.0 is currently in Beta and set to be released in the near future.  Alongside it, will require the usage of Gradle 6.  The tinker gradle plugin does not currently work with these versions.

This PR fixes the issues pertaining to the 4.0 android gradle plugin, as well as Gradle 6.  These changes do not break support for the current 3.x series of the android gradle plugin.

### CODE REVIEW

File: third-party/aosp-dexutils/src/main/java/com/tencent/tinker/android/dex/Mutf8.java [hunk 1]
```diff
@@ -107,7 +107,7 @@ public static void encode(byte[] dst, int offset, String s) {
      * Returns an array containing the <i>modified UTF-8</i> form of {@code s}.
      */
     public static byte[] encode(String s) throws UTFDataFormatException {
-        int utfCount = (int) countBytes(s, true);
+        int utfCount = (int) countBytes(s, false);
```
Comment (fmoda3) [inline third-party/aosp-dexutils/src/main/java/com/tencent/tinker/android/dex/Mutf8.java]: This change (and the one below) is due to an interesting issue regarding String lengths.

With the new Android Gradle Plugin 4.0, the dex classes were suddenly containing some `String` values greater than `65535` in length.  When these functions are set to `true`, `countBytes` throws an error if the length is greater than `65535`.  This length, however, I believe is only a problem in Java, which does not let you store `String` literals in the source code greater than `65535`.

This does not appear to be true in Dex, as the Dex format indicates that the size of a `String` is represented by a `uleb128`, which is unbounded.  Essentially, setting this to false, fixed the issue, and the patch generation succeeded.

File: tinker-sample-android/app/build.gradle [hunk 2]
```diff
@@ -86,6 +86,7 @@ android {
          * you can use multiDex and install it in your ApplicationLifeCycle implement
          */
         multiDexEnabled true
+        multiDexKeepProguard file("tinker_multidexkeep.pro")
```
Comment (fmoda3) [inline tinker-sample-android/app/build.gradle]: I found in testing, that without this proguard file (which is recommended by tinker anyway), sometimes the Tinker code would move out of the first dex and cause the patch generation to fail.  Adding this made testing the patch consistent.

File: tinker-sample-android/build.gradle [hunk 3]
```diff
@@ -7,7 +7,7 @@ buildscript {
     }
     dependencies {
         if (project.hasProperty('GRADLE_3') && GRADLE_3.equalsIgnoreCase('TRUE')) {
-            classpath 'com.android.tools.build:gradle:3.2.1'
+            classpath 'com.android.tools.build:gradle:4.0.0-alpha09'
```
Comment (fmoda3) [inline tinker-sample-android/build.gradle]: This doesn't necessarily need to stay as 4.0.0, but I changed it for testing purposes.

File: tinker-sample-android/gradle.properties [hunk 4]
```diff
@@ -20,5 +20,4 @@ org.gradle.jvmargs=-Xmx2048m -XX:MaxPermSize=1024m -XX:+HeapDumpOnOutOfMemoryErr
 TINKER_VERSION=1.9.14.3
 GRADLE_3=true
 
-android.enableAapt2=false
```
Comment (fmoda3) [inline tinker-sample-android/gradle.properties]: Deprecated in the newer versions

File: tinker-sample-android/gradle/wrapper/gradle-wrapper.properties [hunk 5]
```diff
@@ -1,6 +1,5 @@
-#Tue Oct 24 15:01:44 CST 2017
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.1.1-all.zip
```
Comment (fmoda3) [inline tinker-sample-android/gradle/wrapper/gradle-wrapper.properties]: Android Gradle Plugin 4.0 requires Gradle 6

**Comment by tomystang** [comment]
Thanks for your PRs with detailed description and advice.

**Pull request merged**